File: crx_kinematics_py/crx_kinematics_py/robot.py
from dataclasses import dataclass

import numpy as np

from crx_kinematics.utils.geometry import (
    construct_plane,
    find_third_triangle_corner,
    find_zeros,
    get_dual_ik_solution,
    harmonize_towards_zero,
    isometry_inv,
    normalized,
)


class CircleEvaluation:
    def __init__(self, q, T_R0_tool, dh_params, O5):
        self.q = q

        r4 = dh_params[3].r
        r5 = dh_params[4].r
        r6 = dh_params[5].r
        a3 = dh_params[2].a

        self.O4 = (T_R0_tool @ [r5 * np.cos(q), r5 * np.sin(q), r6, 1])[:3]
        self.dist_O4 = np.linalg.norm(self.O4)
        self.eq_15_holds = self.dist_O4 <= np.abs(r4) + np.abs(a3)
        # if not self.eq_15_holds:
        #     return

        # Create O0O3O4 plane on which O3 will lie (it passes through [0,0,1])
        self.T_R0_plane = construct_plane(self.O4, np.array([0, 0, 1]))

        # Find O3 in that plane.
        O3UP, O3DOWN = find_third_triangle_corner(AB=self.dist_O4, AC=a3, BC=-r4)

        # Express the O3 candidates in the reference frame
        self.O3UP = self.T_R0_plane[:3, :3] @ [*O3UP, 0]
        self.O3DOWN = self.T_R0_plane[:3, :3] @ [*O3DOWN, 0]

        self.Z5fulllength = O5 - self.O4
        self.Z5 = normalized(self.Z5fulllength)
        self.Z4UP = normalized(self.O4 - self.O3UP)
        self.Z4DOWN = normalized(self.O4 - self.O3DOWN)
        self.dot_product_up = self.Z4UP @ self.Z5
        self.dot_product_down = self.Z4DOWN @ self.Z5

    def determine_joint_values(self, dh_params, O5, T_R0_tool, is_up):
        O6 = T_R0_tool[:3, 3]

        J1 = np.arctan2(self.O4[1], self.O4[0])

        T_L1_L0 = isometry_inv(dh_params[0].T(J1))
        R_L1_L0 = T_L1_L0[:3, :3]
        O_1_3 = R_L1_L0 @ (self.O3UP if is_up else self.O3DOWN)
        J2 = -np.arctan2(O_1_3[2], O_1_3[0]) + np.pi / 2

        O_1_4 = R_L1_L0 @ self.O4
        J3 = np.arctan2(O_1_4[2] - O_1_3[2], O_1_4[0] - O_1_3[0])

        T_L2_L1 = isometry_inv(dh_params[1].T(J2))
        T_L3_L2 = isometry_inv(dh_params[2].T(J3 + J2))  # Account for "Fanuc pecularity"
        T_L3_L0 = T_L3_L2 @ T_L2_L1 @ T_L1_L0
        O_3_5 = T_L3_L0 @ [*O5, 1.0]
        J4 = np.arctan2(O_3_5[0], O_3_5[2])

        T_L4_L3 = isometry_inv(dh_params[3].T(J4))
        T_L4_L0 = T_L4_L3 @ T_L3_L0
        O_4_6 = T_L4_L0 @ [*O6, 1.0]
        J5 = np.arctan2(O_4_6[0], -O_4_6[2])

        T_L5_L0 = isometry_inv(dh_params[4].T(J5)) @ T_L4_L0
        T_L5_tool = T_L5_L0 @ T_R0_tool
        J6 = np.arctan2(-T_L5_tool[2, 0], T_L5_tool[0, 0])

        solution = [np.degrees(J) for J in [J1, J2, J3, J4, J5, J6]]
        return solution


@dataclass
class IKDebugData:
    circle_evaluations: list[CircleEvaluation]
    sample_signal_up: list[float]
    sample_signal_down: list[float]
    up_zeros: list[float]
    down_zeros: list[float]


class CRXRobot:
    @dataclass
    class DHParams:
        a: float = 0.0
        alpha: float = 0.0
        r: float = 0.0
        theta: float = 0.0

        def T(self, joint_value=0):  # in radians
            a, alpha, r, theta = self.a, self.alpha, self.r, self.theta

            theta = joint_value + self.theta

            stheta, ctheta = np.sin(theta), np.cos(theta)
            salpha, calpha = np.sin(alpha), np.cos(alpha)

            T = np.eye(4)  # Assign each row. See Eq 1.
            T[0, :4] = [ctheta, -stheta, 0, a]
            T[1, :4] = [stheta * calpha, ctheta * calpha, -salpha, -r * salpha]
            T[2, :4] = [stheta * salpha, ctheta * salpha, calpha, r * calpha]
            return T

    # Corresponds to each column in Table 2.
    # Note that some rows annoyingly attribute values to i-1 while some attribute to i.
    # The below params attribute all values in each column to i.
    L1 = DHParams()
    L2 = DHParams(alpha=-np.pi / 2, theta=-np.pi / 2)
    L3 = DHParams(a=0.54, alpha=np.pi)
    L4 = DHParams(alpha=-np.pi / 2, r=-0.54)
    L5 = DHParams(alpha=np.pi / 2, r=0.15)
    L6 = DHParams(alpha=-np.pi / 2, r=-0.16)

    dh_params = [L1, L2, L3, L4, L5, L6]

    T_L6_tool = np.diag([1.0, -1, -1, 1])
    frame_names = ["R0", "L1", "L2", "L3", "L4", "L5", "L6", "tool"]

    def __init__(self):
        pass

    def fk(self, joint_values=None, return_individual_transforms=False):
        # Joint values in degrees

        if num_values := len(joint_values) != 6:
            raise RuntimeError(f"Received {len(num_values)} joint values for FK, but expected 6")

        if joint_values is None:
            joint_values = [0] * 6

        joint_values = [np.radians(j) for j in joint_values]  # Express in radians

        joint_values[2] += joint_values[1]  # Account for "Fanuc pecularity" mentioned after Table 2

        T_list = [param.T(j) for j, param in zip(joint_values, self.dh_params)]
        T_list.append(self.T_L6_tool)

        T_R0_tool = T_list[0].copy()
        for T in T_list[1:]:
            T_R0_tool = T_R0_tool @ T

        if return_individual_transforms:
            return T_R0_tool, T_list

        return T_R0_tool

    IKSolution = list[float]

    def ik(self, T_R0_tool) -> tuple[list[IKSolution], IKDebugData]:
        # Step 1: Positioning the centers O6 and O5 in the frame R0
        O6 = T_R0_tool[:3, 3]
        O5 = (T_R0_tool @ [0, 0, self.L6.r, 1])[:3]

        # Step 2: Position all O4 candidates on a circle. Each candidate is a CircleEvaluation
        circle_evaluations = [
            CircleEvaluation(q, T_R0_tool, self.dh_params, O5)
            for q in np.linspace(0, 2 * np.pi, 360)
        ]
        # Each CircleEvaluation also performs
        #  - Step 3: Position the 2 O3 candidates (O3UP, O3DOWN) in the vertical O0O3O4 plane
        #  - Calculate the dot products Z4Z5 for UP and DOWN

        # Step 4: Calculate the sample functions (dot_product_up, dot_product_down)
        sample_signal_up, sample_signal_down, sample_signal_x = zip(
            *[(ce.dot_product_up, ce.dot_product_down, ce.q) for ce in circle_evaluations]
        )

        # Step 5: Find the Ns zeros of the sample functions
        def sample_fn_eval(q, is_up):
            ce = CircleEvaluation(q, T_R0_tool, self.dh_params, O5)
            return ce.dot_product_up if is_up else ce.dot_product_down

        up_zeros, down_zeros = find_zeros(
            sample_signal_up, sample_signal_down, sample_signal_x, fn=sample_fn_eval
        )

        # Step 6: Position O4 and O3, determine joint values geometrically
        ik_sols = [
            CircleEvaluation(root, T_R0_tool, self.dh_params, O5).determine_joint_values(
                self.dh_params, O5, T_R0_tool, is_up
            )
            for root, is_up in zip(
                up_zeros + down_zeros, [True] * len(up_zeros) + [False] * len(down_zeros)
            )
        ]

        # Step 7: Calculate dual solutions for each of the above IK solutions
        ik_sols.extend([get_dual_ik_solution(ik_sol) for ik_sol in ik_sols])

        ik_sols = [harmonize_towards_zero(solution) for solution in ik_sols]

        debug_data = IKDebugData(
            circle_evaluations,
            sample_signal_up,
            sample_signal_down,
            up_zeros,
            down_zeros,
        )

        return ik_sols, debug_data

File: crx_kinematics/src/robot.cpp
#include "crx_kinematics/robot.hpp"
#include <cmath>
#include <iostream>
#include <fstream>
#include <limits>

namespace crx_kinematics
{
namespace
{

/**
 * @brief Corresponds to each column in Table 2 of Abbes and Poisson (2024).
 * Note that some rows annoyingly attribute values to i-1 while some attribute to i.
 * The below params attribute all values in each column to i for convenience, although it's
 * technically incorrect according to the modified DH convention.
 */
std::array<DHParams, 6> crx_10ia_params()
{
    DHParams L1 = {};
    DHParams L2 = { .alpha = -M_PI / 2, .theta = -M_PI / 2 };
    DHParams L3 = { .a = 0.54, .alpha = M_PI };
    DHParams L4 = { .alpha = -M_PI / 2, .r = -0.54 };
    DHParams L5 = { .alpha = M_PI / 2, .r = 0.15 };
    DHParams L6 = { .alpha = -M_PI / 2, .r = -0.16 };

    return { L1, L2, L3, L4, L5, L6 };
}

const Eigen::Matrix4d T_J6_tool = []() {
    return Eigen::Vector4d(1.0, -1, -1, 1).asDiagonal().toDenseMatrix();
}();

/**
 * @brief Returns the vector rejection of a from b.
 * In other words, the orthogonal component of a that is perpendicular to b.
 * See https://en.wikipedia.org/wiki/Vector_projection
 */
Eigen::Vector3d vector_rejection(const Eigen::Vector3d& a, const Eigen::Vector3d& b)
{
    const Eigen::Vector3d projection = a.dot(b) / b.dot(b) * b;
    return a - projection;
}

/**
 * @brief Construct the plane O0O3O4
 * See Step 3 of Abbes and Poisson (2024)
 */
Eigen::Isometry3d construct_plane(const Eigen::Vector3d& dir1, const Eigen::Vector3d& dir2)
{
    const Eigen::Vector3d x = dir1.normalized();
    const Eigen::Vector3d y = vector_rejection(dir2, dir1).normalized();
    const Eigen::Vector3d z = x.cross(y);

    Eigen::Matrix3d R = Eigen::Matrix3d();
    R << x, y, z;

    Eigen::Isometry3d out;
    out.translation().setZero();
    out.linear() = R;
    return out;
}

/**
 * @brief Given a triangle where
      1. The lengths of each side are known,
      2. The first corner is at (0, 0)
      3. The second corner is at (AB, 0)
    Returns the position (x, y) of the third triangle.
    Note that flipping the sign of y also is a valid solution, hence both solutions are returned

           C = (x, y)
         /   \
      /        \
    A ---------- B = (AB, 0)

    https://math.stackexchange.com/a/544025
 */
std::pair<Eigen::Vector2d, Eigen::Vector2d> find_third_triangle_corner(const double AB,
                                                                       const double AC,
                                                                       const double BC)
{
    const double x = (AC * AC - BC * BC + AB * AB) / (2 * AB);

    if (x * x > AC * AC)
    {
        // Triangle inequality doesn't hold.
        return std::make_pair(Eigen::Vector2d(x, 0), Eigen::Vector2d(x, 0));
    }
    const double y = std::sqrt((AC * AC) - (x * x));

    return std::make_pair(Eigen::Vector2d(x, y), Eigen::Vector2d(x, -y));
}

/**
 * @brief Finds the root of a real function f given an interval [a, b] where f(a) and f(b) have
 * different signs. This function implements the bisection method:
 * https://en.wikipedia.org/wiki/Bisection_method
 * @param a One side of the interval
 * @param b The other side of the interval
 * @param f The function to find the root of. In our context, this function takes a double q and
 * returns one of the "dot product" values of a CircleEvaluation (corresponding to O3UP or O3DOWN)
 * @param fa The function value evaluated at a
 * @param fb The function value evaluated at b
 * @param tolerance Dictates when a found q is deemed acceptably close to zero.
 * @param max_iterations Dictates the max number of iterations to perform in case a acceptable
 * solution is not found
 * @return The found root of the function
 */
double find_zero(double a,
                 double b,
                 std::function<double(double)> f,
                 double fa,
                 double fb,
                 const double tolerance = 1e-6,
                 const int max_iterations = 100)
{
    if (std::fabs(fa) < tolerance)
    {
        return a;
    }
    if (std::fabs(fb) < tolerance)
    {
        return b;
    }

    if (fa < fb)
    {
        std::swap(a, b);
        std::swap(fa, fb);
    }

    double q = 0.5 * (a + b);
    double dot_product = f(q);

    int num_iterations = 1;

    while (std::fabs(dot_product) >= tolerance)
    {
        if (num_iterations == max_iterations)
        {
            break;
        }

        if (dot_product >= 0.0)
        {
            a = q;
            fa = dot_product;
        }
        else
        {
            b = q;
            fb = dot_product;
        }

        q = 0.5 * (a + b);
        dot_product = f(q);

        ++num_iterations;
    }

    return q;
}

/**
 * @brief Given valid placements of the origins of J3 and J4, the values of each joint can be
 * computed. This function does that.
 * See Step 6 of Abbes and Poisson 2024.
 * @return The IK solution for this configuration of O3 and O4.
 */
std::array<double, 6> determine_joint_values(const Eigen::Vector3d& O3,
                                             const Eigen::Vector3d& O4,
                                             const Eigen::Vector3d& O5,
                                             const Eigen::Isometry3d& T_R0_tool,
                                             const std::array<DHParams, 6>& dh_params)
{
    const auto O6 = T_R0_tool.translation();
    const double J1 = std::atan2(O4.y(), O4.x());

    const auto T_L1_L0 = dh_params[0].T(J1).inverse();
    const auto R_L1_L0 = T_L1_L0.linear();
    const auto O_1_3 = R_L1_L0 * O3;
    const double J2 = std::atan2(O_1_3.x(), O_1_3.z());

    const auto O_1_4 = R_L1_L0 * O4;
    const double J3 = std::atan2(O_1_4.z() - O_1_3.z(), O_1_4.x() - O_1_3.x());

    const auto T_L2_L1 = dh_params[1].T(J2).inverse();
    const auto T_L3_L2 = dh_params[2].T(J2 + J3).inverse();  // Handle Fanuc J2/J3 coupling

    const auto T_L3_L0 = T_L3_L2 * T_L2_L1 * T_L1_L0;
    const auto O_3_5 = T_L3_L0 * O5;
    const double J4 = std::atan2(O_3_5.x(), O_3_5.z());

    const auto T_L4_L3 = dh_params[3].T(J4).inverse();
    const auto T_L4_L0 = T_L4_L3 * T_L3_L0;
    const auto O_4_6 = T_L4_L0 * O6;
    const double J5 = std::atan2(O_4_6.x(), -O_4_6.z());

    const auto T_L5_L0 = dh_params[4].T(J5).inverse() * T_L4_L0;
    const auto T_L5_tool = T_L5_L0 * T_R0_tool;
    const auto tool_x_in_L5 = T_L5_tool.linear().col(0);
    const double J6 = std::atan2(-tool_x_in_L5.z(), tool_x_in_L5.x());

    return { J1, J2, J3, J4, J5, J6 };
}

double harmonize_towards_zero(const double x)
{
    if (x < -M_PI)
    {
        return x + 2.0 * M_PI;
    }
    if (x > M_PI)
    {
        return x - 2.0 * M_PI;
    }
    return x;
};

}  // namespace

CircleEvaluation::CircleEvaluation(const double q,
                                   const Eigen::Isometry3d& T_R0_tool,
                                   const double r4,
                                   const double r5,
                                   const double r6,
                                   const double a3,
                                   const Eigen::Vector3d& O5)
  : q(q)
{
    O4 = T_R0_tool * Eigen::Vector3d(r5 * std::cos(q), r5 * std::sin(q), r6);
    triangle_inequality_holds = O4.norm() <= std::abs(r4) + std::abs(a3);

    const Eigen::Isometry3d T_R0_plane = construct_plane(O4, Eigen::Vector3d::UnitZ());
    const auto [corner_up, corner_down] = find_third_triangle_corner(O4.norm(), a3, -r4);

    O3UP = T_R0_plane * Eigen::Vector3d(corner_up.x(), corner_up.y(), 0.0);
    O3DOWN = T_R0_plane * Eigen::Vector3d(corner_down.x(), corner_down.y(), 0.0);

    const Eigen::Vector3d Z5 = (O5 - O4).normalized();
    dot_product_up = Eigen::Vector3d(O4 - O3UP).normalized().dot(Z5);
    dot_product_down = Eigen::Vector3d(O4 - O3DOWN).normalized().dot(Z5);
}

Eigen::Isometry3d DHParams::T(double joint_value) const
{
    double stheta = std::sin(joint_value + theta);
    double ctheta = std::cos(joint_value + theta);
    double salpha = std::sin(alpha);
    double calpha = std::cos(alpha);

    Eigen::Matrix4d T_parent_child;
    // Matching Eq 1. of Abbes and Poisson 2024
    T_parent_child << ctheta, -stheta, 0, a,                     //
        stheta * calpha, ctheta * calpha, -salpha, -r * salpha,  //
        stheta * salpha, ctheta * salpha, calpha, r * calpha,    //
        0, 0, 0, 1;

    Eigen::Isometry3d out;
    out.matrix() = T_parent_child;
    return out;
}

CRXRobot::CRXRobot() : dh_params(crx_10ia_params())
{
}

Eigen::Isometry3d CRXRobot::fk(const std::array<double, 6>& joint_values) const
{
    Eigen::Isometry3d out = dh_params[0].T(joint_values[0]);
    out = out * dh_params[1].T(joint_values[1]);
    out = out * dh_params[2].T(joint_values[2] + joint_values[1]);  // Handle Fanuc J2/J3 coupling
    out = out * dh_params[3].T(joint_values[3]);
    out = out * dh_params[4].T(joint_values[4]);
    out = out * dh_params[5].T(joint_values[5]);
    out = out * T_J6_tool;
    return out;
}

std::vector<std::array<double, 6>> CRXRobot::ik(const Eigen::Isometry3d& desired_pose) const
{
    // Step 1
    const Eigen::Vector3d O5 =
        desired_pose /*aka T_R0_tool*/ * Eigen::Vector3d(0.0, 0.0, dh_params[5].r);

    const double r4 = dh_params[3].r;
    const double r5 = dh_params[4].r;
    const double r6 = dh_params[5].r;
    const double a3 = dh_params[2].a;

    auto make_circle_evaluation = [&desired_pose, r4, r5, r6, a3, &O5](double q) {
        return CircleEvaluation(q, desired_pose, r4, r5, r6, a3, O5);
    };

    auto f_up = [&make_circle_evaluation](double q) {
        return make_circle_evaluation(q).dot_product_up;
    };
    auto f_down = [&make_circle_evaluation](double q) {
        return make_circle_evaluation(q).dot_product_down;
    };

    auto [previous_up_dot, previous_down_dot] = [&make_circle_evaluation]() {
        const auto ce = make_circle_evaluation(0.0);
        return std::make_pair(ce.dot_product_up, ce.dot_product_down);
    }();

    // Step 2, 3, 4, 5 and 6
    std::vector<std::array<double, 6>> solutions;
    solutions.reserve(16);
    for (int q_deg = 1; q_deg < 360; ++q_deg)
    {
        auto circle_evaluation = CircleEvaluation(
            static_cast<double>(q_deg) / 180.0 * M_PI, desired_pose, r4, r5, r6, a3, O5);

        const double up_dot = circle_evaluation.dot_product_up;
        const double down_dot = circle_evaluation.dot_product_down;

        if (std::signbit(up_dot) != std::signbit(previous_up_dot))
        {
            const double root = find_zero(static_cast<double>(q_deg - 1) / 180.0 * M_PI,
                                          static_cast<double>(q_deg) / 180.0 * M_PI,
                                          f_up,
                                          previous_up_dot,
                                          up_dot);
            const CircleEvaluation ce = CircleEvaluation(root, desired_pose, r4, r5, r6, a3, O5);
            if (ce.triangle_inequality_holds)
            {
                solutions.push_back(
                    determine_joint_values(ce.O3UP, ce.O4, O5, desired_pose, dh_params));
            }
        }
        if (std::signbit(down_dot) != std::signbit(previous_down_dot))
        {
            const double root = find_zero(static_cast<double>(q_deg - 1) / 180.0 * M_PI,
                                          static_cast<double>(q_deg) / 180.0 * M_PI,
                                          f_down,
                                          previous_down_dot,
                                          down_dot);
            const CircleEvaluation ce = CircleEvaluation(root, desired_pose, r4, r5, r6, a3, O5);
            if (ce.triangle_inequality_holds)
            {
                solutions.push_back(
                    determine_joint_values(ce.O3DOWN, ce.O4, O5, desired_pose, dh_params));
            }
        }

        previous_up_dot = up_dot;
        previous_down_dot = down_dot;
    }

    // Step 7
    const auto num_regular_solutions = solutions.size();
    for (auto i = 0u; i < num_regular_solutions; ++i)
    {
        const auto& sol = solutions[i];
        std::array<double, 6> dual = {
            harmonize_towards_zero(sol[0] - M_PI),
            harmonize_towards_zero(-sol[1]),
            harmonize_towards_zero(M_PI - sol[2]),
            harmonize_towards_zero(sol[3] - M_PI),
            sol[4],
            sol[5],
        };
        solutions.push_back(dual);
    }

    return solutions;
}

std::array<double, 6> to_xyzwpr(const Eigen::Isometry3d& T)
{
    const Eigen::Vector3d xyz = T.translation() * 1000.0;

    const Eigen::Matrix3d R = T.linear();
    // https://github.com/matthew-brett/transforms3d/blob/dc877e1/transforms3d/euler.py#L237
    const double cy = std::sqrt(R(0, 0) * R(0, 0) + R(1, 0) * R(1, 0));
    Eigen::Vector3d wpr{};
    if (cy > std::numeric_limits<double>::epsilon() * 4.0)
    {
        wpr = Eigen::Vector3d(std::atan2(R(2, 1), R(2, 2)) * 180.0 / M_PI,
                              std::atan2(-R(2, 0), cy) * 180.0 / M_PI,
                              std::atan2(R(1, 0), R(0, 0)) * 180.0 / M_PI);
    }
    else
    {
        wpr = Eigen::Vector3d(std::atan2(-R(1, 2), R(1, 1)) * 180.0 / M_PI,
                              std::atan2(-R(2, 0), cy) * 180.0 / M_PI,
                              0.0);
    }

    return { xyz[0], xyz[1], xyz[2], wpr[0], wpr[1], wpr[2] };
}

Eigen::Isometry3d from_xyzwpr(const std::array<double, 6>& xyzwpr)
{
    // xyzwpr expressed in millimeters and degrees, for convenience
    Eigen::Isometry3d T;
    T.translation().x() = xyzwpr[0] / 1000.0;
    T.translation().y() = xyzwpr[1] / 1000.0;
    T.translation().z() = xyzwpr[2] / 1000.0;

    Eigen::Quaterniond rotation =
        Eigen::AngleAxisd(xyzwpr[5] / 180.0 * M_PI, Eigen::Vector3d::UnitZ()) *
        Eigen::AngleAxisd(xyzwpr[4] / 180.0 * M_PI, Eigen::Vector3d::UnitY()) *
        Eigen::AngleAxisd(xyzwpr[3] / 180.0 * M_PI, Eigen::Vector3d::UnitX());

    T.linear() = Eigen::Matrix3d(rotation);
    return T;
}

std::array<double, 6> deg2rad(const std::array<double, 6>& joint_values)
{
    std::array<double, 6> out{};
    for (auto i = 0u; i < joint_values.size(); ++i)
    {
        out[i] = joint_values[i] / 180.0 * M_PI;
    }
    return out;
};

}  // namespace crx_kinematics
